oxd-server/src/main/java/org/xdi/oxd/server/op/GetTokensByCodeOperation.java

```diff
@@ -61,44 +72,67 @@ public IOpResponse execute(GetTokensByCodeParams params) throws Exception {
 
         if (response.getStatus() == 200 || response.getStatus() == 302) { // success or redirect
 
-            if (Strings.isNullOrEmpty(response.getIdToken())) {
-                LOG.error("id_token is not returned. Please check whether 'openid' scope is present for 'get_authorization_url' command");
-                throw new HttpException(ErrorResponseCode.NO_ID_TOKEN_RETURNED);
-            }
+            return handleSuccess(params, site, discoveryResponse, response);
```

Comment:
lets remove all these `if`, it's hard to read such code. Simply do 
```
public void validateResponse(TokenResponse response) {
    if (response.getStatus() == 400) {
        throw new HttpException(INVALID_AUTHORIZATION_CODE_BAD_CODE);
    }
}
```

then in execute it's just
```
 @Override
     public IOpResponse execute(GetTokensByCodeParams params) throws Exception {	     public IOpResponse execute(GetTokensByCodeParams params) throws Exception {
          ...
          validateResponse(tokenResponse)
          ...

          final Jwt idToken = Jwt.parse(response.getIdToken());
          ...
     }
```

At the end it will be liner without all those `handleBadRequest`, `handleSuccess` and so on.

Reply:
We don't want spaghetti code here, like handle success during validation. Simply revert to original code and add `public void validateResponse(TokenResponse response)` which return void. Validation just validates (throws exceptions if something goes wrong). See my example above.